enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,213 @@
+IPI workflow installs, so that the customer have the flexility to configure their own loadbalancer with the ingress and API VIPs pointing to nodes in the separate subnets and manage how the VIPs are routed how they want (e.g. routed via BGP).
+
+### User Stories
+
+As a deployer of a IPI OpenShift cluster, I want to place my control plane nodes and compute nodes on 3 or
+more subnets at installation time. Therefore I need to be able to configure an external load balancer for
+the VIPs.
+
+### Goals
+
+Allow to replace the internal loadbalancer by an external one and support this scenario when the customer
+know what they're doing.
+
+### Non-Goals
+
+* Making this scenario by default at some point. The internal loadbalancer for on-premise platforms will
+  remain the HAproxy/Keepalived/CoreDNS stack.
+* Supporting a mix of both (internal + external loadbalancers for the API & Ingress VIPs), customers
+  will have to choose a scenario.
+
+## Proposal
+
+We will make some change in `install-config.yaml` so customer can configure the
+deployment to use an external loadbalancer if they want.
+Then the components which interact with networking will have to check this
+new field and handle this scenario appropriately.
+
+### Workflow Description
+
+1. The customer will configure their external loadbalancer(s), which includes configuring the VIPs
+   for API and Ingress.
+2. The customer will populate their external DNS server with required records for api, api-int, and ingress.
+3. The customer will write the `install-config.yaml` file and set the new `loadBalancerType` field to `external`, and
+   will not set the fields for apiVIP and ingressVIP. We should fail if they set these mutually exclusive parameters.
+4. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). A new field in `PlatformStatus` will be set.
+5. In machine-config-operator, if LB type is external, the internal loadbalancer will not be
+   deployed.
+6. When using an external loadbalancer, the VIPs fields won't be set in `PlatformStatus` and the
+   OpenShift components won't use it when they're not present and LB type is external.
+7. After the deployment, the IP address of a service (e.g. API) can change, as long as the external
+   loadbalancer is updated and the DNS record has been updated and the results are propagated as well.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// LoadBalancerType can be either external or internal 
+type LoadBalancerType string
+const (
+	// LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "external"
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "internal"
```

Comment:
"internal" is  a specific, technical term for load balancers on several cloud platforms. When we say "internal" we are normally referring to load balancers that are only accessible on the cluster subnets. I think the intent for these values is to represent whether they are customer-managed. Are there other values besides "external"/"internal" that could be used so there is less overlap in terminology?

Reply:
+1 using a more specific term might be beneficial, `CustomerManaged`/`OpenShiftManaged` maybe? Whatever we choose, it should be `PascalCase` btw as this is he convention for enums in Kube

Reply:
> Good suggestions, thanks. Can I still keep `LoadBalancerTypeExternal` and `LoadBalancerTypeInternal` or should also it be renamed to something?

Personally, I would prefer the variable name be consistent (to some degree) with the value.

On a general note: my understanding is that this enhancement is intended for on-prem platforms but could also apply to cloud platforms. In gcp, azure, and aws we always deploy both an external and internal load balancer (to provide split-horizon dns), so the choice between "external" and "internal" does not map well to those platforms. I hope that provides more context to my comment.

Reply:
> Personally, I would prefer the variable name be consistent (to some degree) with the value.

Me too. I'll probably ask @deads2k his feedback and maybe some naming proposals so I can update my API PR.